File: data structures and algorithms/java/08_022.java
import java.util.HashMap;
import java.util.Map;

/**
 * Scheduler for tasks with cooldown period
 */
public class TaskScheduler {
   /**
    * Calculate minimum time to finish all tasks with cooldown
    * 
    * @param tasks Array of tasks to be executed
    * @param n Cooldown period between identical tasks
    * @return Minimum time required to complete all tasks
    */
   public static int minTimeToFinishTasks(char[] tasks, int n) {
      if (tasks == null || tasks.length == 0) {
         return 0;
      }
      
      // Count frequency of each task
      Map<Character, Integer> taskCounts = new HashMap<>();
      for (char task : tasks) {
         taskCounts.put(task, 
                     taskCounts.getOrDefault(task, 0) + 1);
      }
      
      // Find task with maximum frequency
      int maxFreq = 0;
      for (int count : taskCounts.values()) {
         maxFreq = Math.max(maxFreq, count);
      }
      
      // Count tasks with maximum frequency
      int maxFreqTasks = 0;
      for (int count : taskCounts.values()) {
         if (count == maxFreq) {
            maxFreqTasks++;
         }
      }
      
      // Calculate using the formula:
      // (maxFreq-1) * (n+1) + maxFreqTasks
      int result = (maxFreq - 1) * (n + 1) + maxFreqTasks;
      
      // Result should be at least the total task count
      return Math.max(result, tasks.length);
   }
   
   /**
    * Test the scheduler with example inputs
    */
   public static void main(String[] args) {
      // Example 1
      char[] tasks1 = {'A', 'A', 'B', 'C', 'A', 'B'};
      int n1 = 2;
      System.out.println("Example 1 result: " + 
            minTimeToFinishTasks(tasks1, n1));  // Output: 7
      
      // Example 2
      char[] tasks2 = {'A', 'A', 'A', 'B', 'B', 'B'};
      int n2 = 2;
      System.out.println("Example 2 result: " + 
            minTimeToFinishTasks(tasks2, n2));  // Output: 8
   }
}

File: data structures and algorithms/cs/08_022.cs
using System;
using System.Collections.Generic;

public class TaskScheduler
{
   public static int MinTimeToFinishTasks(
      char[] tasks, int n)
   {
      // Handle edge cases
      if (tasks == null || tasks.Length == 0)
      {
         return 0;
      }

      // Count frequency of each task
      var taskCounts = new Dictionary<char, int>();
      foreach (var task in tasks)
      {
         if (!taskCounts.ContainsKey(task))
         {
            taskCounts[task] = 0;
         }
         taskCounts[task]++;
      }

      // Find the task that appears most frequently
      int maxFreq = 0;
      foreach (var count in taskCounts.Values)
      {
         maxFreq = Math.Max(maxFreq, count);
      }

      // Count how many tasks have the maximum frequency
      int maxFreqTasks = 0;
      foreach (var count in taskCounts.Values)
      {
         if (count == maxFreq)
         {
            maxFreqTasks++;
         }
      }

      // Calculate minimum time required:
      int result = (maxFreq - 1) * (n + 1) + maxFreqTasks;

      // Return the larger of our calculated value 
      // and the total number of tasks
      return Math.Max(result, tasks.Length);
   }

   public static void Main(string[] args)
   {
      char[] tasks1 = { 'A', 'A', 'B', 'C', 'A', 'B' };
      int n1 = 2;
      Console.WriteLine("Example 1 result: " +
         MinTimeToFinishTasks(tasks1, n1));  // Output: 7

      char[] tasks2 = { 'A', 'A', 'A', 'B', 'B', 'B' };
      int n2 = 2;
      Console.WriteLine("Example 2 result: " +
         MinTimeToFinishTasks(tasks2, n2));  // Output: 8
   }
}